esp32_ulp/preprocess.py

```diff
@@ -0,0 +1,53 @@
+from . import nocomment
+from .util import split_tokens
+
+
+class Preprocessor:
+    def __init__(self):
+        self._defines = {}
+
+    def parse_defines(self, content):
+        result = {}
+        for line in content.splitlines():
```

Comment:
avoid by doing nocomment first?

Reply:
Good idea, but nocomment would strip out the #define lines, so we need to parse them first.

One could potentially adapt nocomment to accept a list of lines as input, rather than a single string, so that there isnt this back and forth conversion all the time. I had actually thought of that as a future potential improvement to use even less memory - to make all parts capable of operating line-by-line, either using the generator pattern or by returning state (e.g. currently inside a multiline comment) that gets passed into the next call as the starting state, etc..

But for now I have managed to assemble (including the preprocessing) up to 6k of input code on the actual ESP32 which was more than almost all the example code I found and tested with. When doing "aggressive" garbage collect - in the preprocess main loop after every line, then even the biggest file I had worked.

Do you feel this should be improved or ok to leave as is?

Reply:
maybe file a ticket about refactoring this later, this PR is already way to big, so we avoid such refactorings here. splitting once and avoiding join/split/join/split.

Reply:
Yes. Lets do that. Looks like issue #38 already cover this. Once the preprocess is merged into master, I will update that issue, and add that there's opportunity in the preprocessor for saving memory too.